### PR TITLE
[balsa] Ignore long trailers.

### PR DESCRIPTION
**Pull request by bencebeky**
Ignore trailers with long key or value if
codec_settings_.enable_trailers_ is false.

This also causes trailers not to be validated if
codec_settings_.enable_trailers_ is false, which is different from http-parser behavior.

Tracking issue: #21245

Signed-off-by: Bence Béky <bnc@google.com>

Commit Message: [balsa] Ignore long trailers.
Additional Description: Ignore trailers with long key or value if codec_settings_.enable_trailers_ is false.
Risk Level: low
Testing: //test/common/http/http1:codec_impl_test
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

File: test/common/http/http1/codec_impl_test.cc [hunk 1]
```diff
@@ -928,6 +928,11 @@ TEST_P(Http1ServerConnectionImplTest, Http11InvalidRequest) {
 }
 
 TEST_P(Http1ServerConnectionImplTest, Http11InvalidTrailerPost) {
+  if (parser_impl_ == ParserImpl::BalsaParser) {
+    // BalsaParser only validates trailers if `enable_trailers_` is set.
+    codec_settings_.enable_trailers_ = true;
+  }
```
Comment (diannahu) [inline test/common/http/http1/codec_impl_test.cc:934]: This sounds reasonable to me and indeed seems to be more in the spirit of https://github.com/envoyproxy/envoy/blob/main/source/common/http/http1/codec_impl.cc#L765-L768.

**Review by diannahu — DISMISSED** [review]
Nice! A short change to ignore long trailers :)

**Comment by bencebeky** [comment]
Dianna: Thank you for the review.
Kevin: PTAL

/assign @KBaichoo

**Review by KBaichoo — COMMENTED** [review]
Seems reasonable that we'd not validate trailers if they're not enabled. How does balsa end up cleaning up the buffer with invalid trailers?

**Comment by bencebeky** [comment]
> Seems reasonable that we'd not validate trailers if they're not enabled. How does balsa end up cleaning up the buffer with invalid trailers?

BalsaFrame does not buffer the raw input, instead it copies headers (and trailers, if `trailer_` is set) into BalsaHeaders objects, and calls BalsaVisitorInterface methods with absl::string_view arguments that point into the raw input.

**Comment by KBaichoo** [comment]
Adding wait as discussed out of band for a test that has balsa trailers disabled but invalid trailers, we can recover and process a subsequent well formed request

**Review by KBaichoo — APPROVED** [review]
lgtm, thanks

**Comment by bencebeky** [comment]
Thank you.

**Pull request merged**